docs: add documentation for CVE records (#1182)
Nix-Security-WG/nix-security-tracker: docs/cve_records.md

# CVE records

The tracker parses [CVE JSON 5.0](https://github.com/CVEProject/cve-schema) records from the [official CVE List](https://github.com/CVEProject/cvelistV5) and stores them in the database.
The schema follows the upstream format; see the [CVE record content rules](https://www.cve.org/ResourcesSupport/AllResources/CNARules#section_5_CVE_Record_Content) and the [schema reference](https://cveproject.github.io/cve-schema/schema/docs/) for field definitions.

Implementation: [`src/shared/models/cve.py`](../src/shared/models/cve.py).

> A `CveRecord` has one or more `Container`s (CNA or ADP). Matching mainly uses each container's `AffectedProduct` entries: vendor/product names, package names, CPE strings, and version constraints.

Reviewer: ```suggestion
A `CveRecord` has one or more `Container`s (CNA or ADP).
Matching mainly uses each container's `AffectedProduct` entries: vendor/product names, package names, CPE strings, and version constraints.
```

https://rhodesmill.org/brandon/2012/one-sentence-per-line/

## Why matching needs manual triage

CVE records describe upstream software by vendor names, product names, and CPE strings.
Nixpkgs identifies packages by attribute paths and derivation names.
Those naming schemes rarely align one-to-one, and version information in CVE data is often incomplete or imprecise.

> The tracker [proposes links automatically](../src/shared/listeners/automatic_linkage.py), but a human must verify each match, ignore irrelevant packages, and publish an issue when Nixpkgs is actually affected.

Reviewer: This is conflating two ideas:
1. Why the design is the way it is (goes to the design README)
2. Where the implementation lives and how it works.

It's fine to link from this sentence to a document which talks about the implementation overall, like the upper part of the current file. But the specifics, including the link to the code, should be in that other document IMO.

# CVE records

> The tracker ingests [CVE JSON 5.0](https://github.com/CVEProject/cve-schema) records from the [official CVE List](https://github.com/CVEProject/cvelistV5) and stores them in PostgreSQL. The schema follows the upstream format; see the [CVE record content rules](https://www.cve.org/ResourcesSupport/AllResources/CNARules#section_5_CVE_Record_Content) and the [schema reference](https://cveproject.github.io/cve-schema/schema/docs/) for field definitions.

Reviewer: ```suggestion
The tracker parses [CVE JSON 5.0](https://github.com/CVEProject/cve-schema) records from the [official CVE List](https://github.com/CVEProject/cvelistV5) and stores them in the database.
The schema follows the upstream format; see the [CVE record content rules](https://www.cve.org/ResourcesSupport/AllResources/CNARules#section_5_CVE_Record_Content) and the [schema reference](https://cveproject.github.io/cve-schema/schema/docs/) for field definitions.
```

Implementation: [`src/shared/models/cve.py`](../src/shared/models/cve.py).

> ## Top-level record (`CveRecord`)

Reviewer: I'm very skeptical whether we should duplicate information that already lives in the code. It should be either avoided entirely or have an extraordinarily strong justification. You already linked to the implementation, so why repeat anything that we can find there?

Author: Fair point. The field tables mostly restated the model, and anyone who needs the details can read cve.py or the upstream schema. I'll drop them and keep the doc focused on structure and why triage is still manual.

| Field | Meaning |
| --- | --- |
| `cve_id` | Identifier, E.g. `CVE-2024-1234` |
| `state` | `PUBLISHED` or `REJECTED` |
| `assigner` / `requester` | CNA organizations that assigned or requested the ID |
| `date_published`, `date_updated`, `date_reserved` | Timestamps from the CVE Program |
| `serial` | Record revision number |

> | `triaged` | Internal flag: a security team member has reviewed this CVE |

Reviewer: That's not correct IIRC, it simply says whether it was touched by the system on ingestion (or matching, not sure). But we're not using that for anything. It was originally intended to avoid matching over a CVE twice. The new rematching logic by @adekoder allows displacing that with the fact that a suggestion is linked to it (modulo naming changes proposed by @florentc).

Author: Thanks, I mixed that up, because it is not doing any useful work , I will remove it from the doc

| `problem_types` | CWE identifiers and descriptions |
| `references` | URLs with optional tags (advisory, patch, etc.) |
| `metrics` | CVSS vector strings (v3.0, v3.1, v4.0) |
| `configurations`, `workarounds`, `solutions`, `exploits` | Additional structured text |
| `timeline`, `credits`, `tags` | Events, attribution, labels |
| `source` | Raw JSON fragment from ingestion |

## Affected products (`AffectedProduct`)

This is what the matching algorithm uses to find Nixpkgs packages.

| Field | Meaning |
| --- | --- |
| `vendor`, `product` | Upstream vendor and product name |
| `package_name` | Package name when the CNA provides one |
| `cpes` | [CPE](https://nvd.nist.gov/products/cpe) strings identifying software |
| `versions` | Version constraints with status `affected`, `unaffected`, or `unknown` |
| `platforms` | OS or environment qualifiers |
| `repo`, `collection_url` | Source repository hints |
| `modules`, `program_files`, `program_routines` | Finer-grained location within a product |

Each `Version` entry expresses a constraint: exact version (`version`), upper bound (`less_than`, `less_equal`), and whether that range is affected.

> ## Why matching needs manual triage

Reviewer: This is a good question to answer, but it may be better suited for a more high-level document such as a (to be added) entry point to the design decisions (or a note in one of them), or even the top-level README. But if there's no obvious other place, it's probably fine to leave it here for now.

Author: Let's discuss in sync, for now not moving this.

Reviewer: Let's move it into a new `docs/design/README.md` since this is where the [explanations](https://diataxis.fr/explanation/) live as far as we have them.

> CVE records describe upstream software by vendor names, product names, and CPE strings. Nixpkgs identifies packages by attribute paths and derivation names. Those naming schemes rarely align one-to-one, and version comparison in CVE data is often incomplete or imprecise.

Reviewer: ```suggestion
CVE records describe upstream software by vendor names, product names, and CPE strings.
Nixpkgs identifies packages by attribute paths and derivation names.
Those naming schemes rarely align one-to-one, and version information in CVE data is often incomplete or imprecise.
```

> The tracker proposes links automatically, but a human must verify each match, ignore irrelevant packages, and publish an issue when Nixpkgs is actually affected.

Reviewer: A good opportunity to link to the implementation!
